Error installing cld2 gem with gcc 6+

## Problem

The [current Gemfile](https://github.com/tootsuite/mastodon/blob/4d22d03fabc5f249fd2a93037916b980e4d3f99b/Gemfile#L22) includes `cld2` for language detection functionality, which **seems to be incompatible with gcc-6+**.

I tried to do a simple `bundle install` on my Ubuntu 16.10 machine (which comes with gcc 6.2.0 by default). When processing the `cld2` gem installation (effectively `gem install cld2 -v '1.0.3'`), I got some spectacular error messages:

```
Building native extensions.  This could take a while...
ERROR:  Error installing cld2:
	ERROR: Failed to build gem native extension.

    current directory: /home/foobar/.rvm/gems/ruby-2.4.1/gems/cld2-1.0.3/ext/cld
/home/foobar/.rvm/rubies/ruby-2.4.1/bin/ruby -r ./siteconf20170505-6029-169rn4c.rb extconf.rb
checking for -lstdc++... yes
creating Makefile

current directory: /home/foobar/.rvm/gems/ruby-2.4.1/gems/cld2-1.0.3/ext/cld
make "DESTDIR=" clean

current directory: /home/foobar/.rvm/gems/ruby-2.4.1/gems/cld2-1.0.3/ext/cld
make "DESTDIR="
compiling internal/cldutil.cc
cc1plus: warning: command line option ‘-Wimplicit-int’ is valid for C/ObjC but not for C++
cc1plus: warning: command line option ‘-Wdeclaration-after-statement’ is valid for C/ObjC but not for C++
cc1plus: warning: command line option ‘-Wimplicit-function-declaration’ is valid for C/ObjC but not for C++
In file included from internal/cldutil.h:24:0,
                 from internal/cldutil.cc:20:
internal/cldutil_shared.h:404:54: warning: type qualifiers ignored on function return type [-Wignored-qualifiers]
                                       uint32 quadhash) {
                                                      ^
internal/cldutil_shared.h:432:62: warning: type qualifiers ignored on function return type [-Wignored-qualifiers]
                                           uint64 longwordhash) {
                                                              ^
cc1plus: warning: unrecognized command line option ‘-Wno-self-assign’

...
(skipped)
...

internal/cld_generated_cjk_uni_prop_80.cc:7089:1: error: narrowing conversion of ‘-9’ from ‘int’ to ‘CLD2::uint8 {aka unsigned char}’ inside { } [-Wnarrowing]
internal/cld_generated_cjk_uni_prop_80.cc:7089:1: error: narrowing conversion of ‘-9’ from ‘int’ to ‘CLD2::uint8 {aka unsigned char}’ inside { } [-Wnarrowing]
cc1plus: warning: unrecognized command line option ‘-Wno-self-assign’
cc1plus: warning: unrecognized command line option ‘-Wno-constant-logical-operand’
cc1plus: warning: unrecognized command line option ‘-Wno-parentheses-equality’
Makefile:209: recipe for target 'internal/cld_generated_cjk_uni_prop_80.o' failed
make: *** [internal/cld_generated_cjk_uni_prop_80.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/foobar/.rvm/gems/ruby-2.4.1/gems/cld2-1.0.3 for inspection.
Results logged to /home/foobar/.rvm/gems/ruby-2.4.1/extensions/x86_64-linux/2.4.0/cld2-1.0.3/gem_make.out

```

## Temp Solution (Need Documentation)

I found similar issues on [some gist](https://gist.github.com/application2000/73fd6f4bf1be6600a2cf9f56315a2d91) in Github. Then I tried this(*) before another `bundle install`:

```
sudo apt-get install gcc-5 g++-5 -y
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 60 --slave /usr/bin/g++ g++ /usr/bin/g++-5;
```

Then `bundle install` works.

Probably need to document this somewhere. Just don't know where...

(*) Need to `sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-6 62 --slave /usr/bin/g++ g++ /usr/bin/g++-6` afterwards to revert default gcc version to 6.2.0.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

--- a/.travis.yml
+++ b/.travis.yml
@@ -13,9 +13,9 @@ env:
     - LOCAL_DOMAIN=cb6e6126.ngrok.io
     - LOCAL_HTTPS=true
     - RAILS_ENV=test
-    - CXX=g++-4.8
     - NOKOGIRI_USE_SYSTEM_LIBRARIES=true
     - PARALLEL_TEST_PROCESSORS=2
+    - "PATH=$HOME:$PATH"
 
 addons:
   postgresql: 9.4
@@ -24,8 +24,10 @@ addons:
     - ubuntu-toolchain-r-test
     - trusty-media
     packages:
-    - g++-4.8
     - ffmpeg
+    - g++-6
+    - libprotobuf-dev
+    - protobuf-compiler
 
 rvm:
   - 2.3.4
@@ -43,6 +45,7 @@ install:
 before_script:
   - bundle exec rake parallel:create parallel:load_schema parallel:prepare
   - bundle exec rails assets:precompile
+  - ln -s /usr/bin/x86_64-linux-gnu-g++-6 "$HOME/g++"
 
 script:
   - bundle exec parallel_test spec/ --group-by filesize --type rspec

--- a/Dockerfile
+++ b/Dockerfile
@@ -16,7 +16,8 @@ RUN echo "@edge https://nl.alpinelinux.org/alpine/edge/main" >> /etc/apk/reposit
     libxml2-dev \
     libxslt-dev \
     python \
-    build-base" \
+    build-base \
+    protobuf-dev" \
  && apk -U upgrade && apk add \
     $BUILD_DEPS \
     nodejs@edge \
@@ -29,6 +30,7 @@ RUN echo "@edge https://nl.alpinelinux.org/alpine/edge/main" >> /etc/apk/reposit
     file \
     imagemagick@edge \
     ca-certificates \
+    protobuf \
  && npm install -g npm@3 && npm install -g yarn \
  && update-ca-certificates \
  && rm -rf /tmp/* /var/cache/apk/*

--- a/Gemfile
+++ b/Gemfile
@@ -19,7 +19,7 @@ gem 'paperclip', '~> 5.1'
 gem 'paperclip-av-transcoder'
 
 gem 'addressable'
-gem 'cld2', require: 'cld'
+gem 'cld3', '~> 3.1.0'
 gem 'devise'
 gem 'devise-two-factor'
 gem 'doorkeeper'

--- a/Vagrantfile
+++ b/Vagrantfile
@@ -33,7 +33,9 @@ sudo apt-get install \
   redis-tools \
   postgresql \
   postgresql-contrib \
+  protobuf-compiler \
   yarn \
+  libprotobuf-dev \
   libreadline-dev \
   -y
 

--- a/app/lib/language_detector.rb
+++ b/app/lib/language_detector.rb
@@ -6,6 +6,7 @@ class LanguageDetector
   def initialize(text, account = nil)
     @text = text
     @account = account
+    @identifier = CLD3::NNetLanguageIdentifier.new(1, 2048)
   end
 
   def to_iso_s
@@ -15,15 +16,15 @@ class LanguageDetector
   private
 
   def detected_language_code
-    detected_language[:code].to_sym if detected_language_reliable?
+    result.language.to_sym if detected_language_reliable?
   end
 
-  def detected_language
-    @_detected_language ||= CLD.detect_language(text_without_urls)
+  def result
+    @result ||= @identifier.find_language(text_without_urls)
   end
 
   def detected_language_reliable?
-    detected_language[:reliable]
+    result.reliable?
   end
 
   def text_without_urls

--- a/spec/lib/language_detector_spec.rb
+++ b/spec/lib/language_detector_spec.rb
@@ -24,15 +24,15 @@ describe LanguageDetector do
     end
 
     describe 'when language can\'t be detected' do
-      it 'confirm language engine cant detect' do
-        result = CLD.detect_language('')
-        expect(result[:reliable]).to be false
+      it 'uses default locale when sent an empty document' do
+        result = described_class.new('').to_iso_s
+        expect(result).to eq :en
       end
 
       describe 'because of a URL' do
         it 'uses default locale when sent just a URL' do
           string = 'http://example.com/media/2kFTgOJLXhQf0g2nKB4'
-          cld_result = CLD.detect_language(string)[:code]
+          cld_result = CLD3::NNetLanguageIdentifier.new(0, 2048).find_language(string)
           expect(cld_result).not_to eq :en
 
           result = described_class.new(string).to_iso_s

